src/dbnode/persist/fs/streaming_write.go

```diff
@@ -0,0 +1,259 @@
+	writer       *writer
+	writerOpts   DataWriterOpenOptions
+	data         []checked.Bytes
+	currIdx      int64
+	prevIDBytes  []byte
+	summaryEvery int64
+	bloomFilter  *bloom.BloomFilter
+	indexOffset  int64
+	summaries    int
+}
+
+func NewStreamingWriter(opts StreamingWriterOptions) (StreamingWriter, error) {
+	w, err := NewWriter(opts.Options)
+	if err != nil {
+		return nil, err
+	}
+
+	writerOpts := DataWriterOpenOptions{
+		BlockSize: opts.BlockSize,
+		Identifier: FileSetFileIdentifier{
+			Namespace:   opts.NamespaceID,
+			Shard:       opts.ShardID,
+			BlockStart:  opts.BlockStart,
+			VolumeIndex: opts.VolumeIndex,
+		},
+		FileSetType: persist.FileSetFlushType,
+	}
+
+	m, k := bloom.EstimateFalsePositiveRate(
+		opts.PlannedRecordsCount,
+		opts.Options.IndexBloomFilterFalsePositivePercent(),
+	)
+	bloomFilter := bloom.NewBloomFilter(m, k)
+
+	summariesApprox := float64(opts.PlannedRecordsCount) * opts.Options.IndexSummariesPercent()
+	summaryEvery := 0
+	if summariesApprox > 0 {
+		summaryEvery = int(math.Floor(float64(opts.PlannedRecordsCount) / summariesApprox))
+	}
+
+	return &streamingWriter{
+		opts:         opts,
+		writer:       w.(*writer),
+		writerOpts:   writerOpts,
+		summaryEvery: int64(summaryEvery),
+		data:         make([]checked.Bytes, 2),
+		bloomFilter:  bloomFilter,
+	}, nil
+}
+
+func (w *streamingWriter) Open() error {
+	if err := w.writer.Open(w.writerOpts); err != nil {
+		return err
+	}
+	w.indexOffset = 0
+	w.summaries = 0
+	w.prevIDBytes = nil
+	return nil
+}
+
+func (w *streamingWriter) Write(
+	ctx context.Context,
+	encoder encoding.Encoder,
+	id ident.ID,
+	encodedTags []byte,
+) error {
+	// Need to check if w.prevIDBytes != nil, otherwise we can never write an empty string ID
+	if w.prevIDBytes != nil && bytes.Compare(id.Bytes(), w.prevIDBytes) <= 0 {
+		return fmt.Errorf("ids must be written in lexicographic order, no duplicates, but got %s followed by %s", w.prevIDBytes, id)
+	}
+	w.prevIDBytes = append(w.prevIDBytes[:0], id.Bytes()...)
+
+	stream, ok := encoder.Stream(ctx)
+	if !ok {
+		// None of the datapoints passed the predicate.
+		return nil
+	}
+	defer stream.Finalize()
+	segment, err := stream.Segment()
+	if err != nil {
+		return err
+	}
+	w.data[0] = segment.Head
+	w.data[1] = segment.Tail // FIXME we fail to finalize the tail (does not get finalized automatically because FinalizeTail flag not set)
+	checksum := segment.CalculateChecksum()
+	entry, err := w.writeData(w.data, checksum)
+	if err != nil {
+		return err
+	}
+
+	if entry != nil {
+		return w.writeIndexRelated(id, encodedTags, entry)
+	}
+
+	return nil
+}
+
+func (w *streamingWriter) writeData(
+	data []checked.Bytes,
+	dataChecksum uint32,
+) (*indexEntry, error) {
+	var size int64
+	for _, d := range data {
+		if d == nil {
+			continue
+		}
+		size += int64(d.Len())
+	}
+	if size == 0 {
+		return nil, nil
+	}
+
+	// Warning: metadata is not set here and should not be used anywhere below.
+	entry := &indexEntry{
```

Comment:
Would avoid using a pointer here, just use a struct and return by value so that this isn't a heap allocation?

Reply:
If you need to be able to return `nil` to imply empty, just return a bool instead? i.e. `(indexEntry, bool, error)` as return type and check the bool.

Reply:
👍